include/etcpal/cpp/uuid.h

```diff
@@ -0,0 +1,278 @@
+
+/// \file etcpal/cpp/uuid.h
+/// \brief C++ wrapper and utilities for etcpal/uuid.h
+
+#ifndef ETCPAL_CPP_UUID_H_
+#define ETCPAL_CPP_UUID_H_
+
+#include <cstdint>
+#include <cstring>
+#include <array>
+#include <string>
+#include "etcpal/uuid.h"
+#include "etcpal/cpp/inet.h"
+
+namespace etcpal
+{
+// clang-format off
+/// \defgroup etcpal_cpp_uuid uuid (UUIDs)
+/// \ingroup etcpal_cpp
+/// \brief C++ utilities for the \ref etcpal_uuid module.
+///
+/// Provides a class Uuid for generating and inspecting Universal Unique Identifiers (UUIDs).
+///
+/// Default-constructed Uuid instances will be all-zeros ("null"):
+/// \code
+/// etcpal::Uuid uuid;
+/// EXPECT_TRUE(uuid.IsNull());
+/// \endcode
+///
+/// To generate a specific type of UUID, use the static generator functions:
+/// \code
+/// auto uuid1 = etcpal::Uuid::V1(); // Generate a V1 (MAC + time based) UUID.
+/// auto uuid3 = etcpal::Uuid::V3(namespace_uuid, name_data, name_size); // Generate a V3 (name-based, MD5) UUID.
+/// auto uuid4 = etcpal::Uuid::V4(); // Generate a V4 (random) UUID.k
+/// auto uuid5 = etcpal::Uuid::V5(namespace_uuid, name_data, name_size); // Generate a V5 (name-based, SHA-1)
+/// auto uuid_os = etcpal::Uuid::OsPreferred(); // Generate the UUID type preferred by the underlying OS
+/// auto uuid_dev = etcpal::Uuid::Device("My Device Type", dev_mac_address, 0); // Generate a UUID representing an embedded device
+/// \endcode
+///
+/// **Note:** Uuid::V3(), Uuid::V5() and Uuid::Device() are guaranteed to be implemented on all
+/// systems. Uuid::V1(), Uuid::V4() and Uuid::OsPreferred() may not be available on embedded
+/// devices. If not implemented, these functions will return a null Uuid. For more information, see
+/// the \ref etcpal_uuid module.
+///
+/// You can also convert UUIDs to and from strings:
+/// \code
+/// auto uuid = etcpal::Uuid::FromString("1b0c096d-9baa-476d-866d-4e0cad18f5a4");
+/// std::string uuid_str = uuid.ToString();
+/// \endcode
+// clang-format on
+
+/// \ingroup etcpal_cpp_uuid
+/// \brief A wrapper class for the EtcPal UUID type.
+///
+/// Provides C++ syntactic sugar for working with UUIDs. UUIDs are considered immutable after they
+/// are created, so there is no non-const access to the underlying data as of now.
+class Uuid
+{
+public:
+  /// \brief Constructs a null UUID by default.
+  Uuid() = default;
+  // Note: this constructor is not explicit by design, to allow implicit conversions e.g.
+  //   etcpal::Uuid uuid = etcpal_c_function_that_returns_uuid();
+  constexpr Uuid(const EtcPalUuid& c_uuid) noexcept;
+  Uuid& operator=(const EtcPalUuid& c_uuid) noexcept;
+
+  constexpr const EtcPalUuid& get() const noexcept;
+  std::string ToString() const;
+  bool IsNull() const noexcept;
+
+  static Uuid FromString(const char* uuid_str) noexcept;
+  static Uuid FromString(const std::string& uuid_str) noexcept;
+  static Uuid V1() noexcept;
+  static Uuid V3(const Uuid& ns, const void* name, size_t name_len) noexcept;
```

Comment:
For a c++ interface, it would be nice to have std::string interfaces of this and V5 in addition to the void*. Also is there a reason this is void* instead of char* for a name type?

Reply:
Good point. The interface was inspired by the one in the [RFC](https://tools.ietf.org/html/rfc4122) - see the definition of `uuid_create_md5_from_name()` on page 19.

I *think* the reason for this is that the "name" doesn't necessarily have to be a string. See the description of names [in section 4.3](https://tools.ietf.org/html/rfc4122#section-4.3). For example, you could create a UUID that represents a custom struct you've defined by calling `Uuid::V3(ns, &my_struct, sizeof(my_struct))`. Not positive about this though.

I'm open to making an overload that takes `char *` and/or `std::string`

Reply:
Got it! That makes a lot of sense and is something I wasn't familiar with.